.github/workflows/test.yaml

```diff
@@ -10,125 +10,144 @@ concurrency:
   cancel-in-progress: true
 
 jobs:
-    linting:
-        runs-on: ubuntu-latest
-
-        steps:
-        - uses: actions/checkout@v3
-
-        - name: Setup Python
-          uses: actions/setup-python@v4
-          with:
-              python-version: "3.10"
-
-        - name: Install Dependencies
-          run: |
-            python -m pip install --upgrade pip
-            pip install .[lint]
-
-        - name: Run Black
-          run: black --check .
-
-        - name: Run isort
-          run: isort --check-only .
-
-        - name: Run flake8
-          run: flake8 .
-
-        - name: Run mdformat
-          run: mdformat . --check
-
-    type-check:
-        runs-on: ubuntu-latest
-
-        steps:
-        - uses: actions/checkout@v3
-
-        - name: Setup Python
-          uses: actions/setup-python@v4
-          with:
-              python-version: "3.10"
-
-        - name: Install Dependencies
-          run: |
-            python -m pip install --upgrade pip
-            pip install .[lint,test]
-
-        - name: Run MyPy
-          run: mypy .
-
-    functional:
-        runs-on: ${{ matrix.os }}
-
-        strategy:
-            matrix:
-                os: [ubuntu-latest, macos-latest]   # eventually add `windows-latest`
-                python-version: [3.8, 3.9, "3.10", "3.11"]
-
-        env:
+  detect-changes:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v3
+    - name: Install changed-files action
+      run: npm install umani/changed-files
+
+    - name: Detect Changed Files
+      id: changed_files
+      run: npx changed-files --diff-base origin/main --extensions py
+
+  linting:
+    needs: detect-changes
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v3
+
+    - name: Setup Python
+      uses: actions/setup-python@v2
+      with:
+        python-version: "3.10"
+
+    - name: Install Dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install .[lint]
+
+    - name: Run Black
+      if: steps.detect-changes.outputs.files_changed == 'true'
+      run: black --check .
+    - name: Run isort
+      if: steps.detect-changes.outputs.files_changed == 'true'
+      run: isort --check-only .
+    - name: Run flake8
+      if: steps.detect-changes.outputs.files_changed == 'true'
+      run: flake8 .
+    - name: Run mdformat
```

Comment:
mdformat can always run, that is probably easier.
because even if you change the README or this file (test.yaml), mdformat may have something to do